# Immunoglobulin (Ig) Variable Domain Gene Discovery Workflow
---------
## A. BLAST querying Ig genes from related species against the ferret genome.
1. <ins>**Retrieving Ig genes of related species:**</ins> From [IMGT](https://www.imgt.org/vquest/refseqh.html#:~:text=IG%20%22V-REGION%22%2C%20%22D-REGION%22%2C%20%22J-REGION%22%2C%20%22C-GENE%20exon%22%20sets), I downloaded FASTA files of all (variable domain-encoding) gene types from multiple species. In this case, downloaded IGHV, IGKV, IGLV, IGHJ, IGKJ, IGLJ, IGHD files for human (*Homo sapiens*), dog (*Canis lupis familiaris*), and cat (*Felis catus*). These species were chosen for biomedical and phylogenetic relevance to the ferret (*Mustela putorius furo*).

2. <ins>**Reformatting as BLAST queries:**</ins> To all files of a given gene segment type (e.g. IGHVs and IGKVs are both segment type "V"), I applied `tidy_imgt_queries.py` to produce a single FASTA containing segment specific DNA sequences across species. The result was three multi-species FASTAs, one for all Vs, one for all Ds, and one for all Js. These were now reformatted to be used as BLAST queries.

3. <ins>**Setting up BLAST database:**</ins> I created a [local BLAST database](https://www.ncbi.nlm.nih.gov/books/NBK569841/) from the most recent ferret genome assembly, [ASM1176430v1.1](https://www.ncbi.nlm.nih.gov/datasets/genome/GCF_011764305.1/).

4. <ins>**Performing BLAST search:**</ins> I used [local BLAST](https://blast.ncbi.nlm.nih.gov/doc/blast-help/downloadblastdata.html#downloadblastdata) parameters detailed in `v_blast_query.sh`, `d_blast_query.sh`, and `j_blast_query.sh` to search the ferret genome using reformatted IMGT sequences.

5. <ins>**Keep longest overlapping hits:**</ins> I ran `overlap_filter.py` on each of the resulting BLAST hit TSV files to apply a sweep-line algorithm that keeps only the longest BLAST hit for any overlapping group of hits. 
## B. Annotating Ig gene features.
1. <ins>**Gathering flanking sequences:**</ins> I applied `flanking_seqs.py` to non-overlapping BLAST hits to annotate them with regions that would have recombination signal sequences (RSSs) and leader peptides (for V genes), if the hits are Ig genes.

2. <ins>**Evaluating flanking sequences for RSS candidates:**</ins> Results from the previous step, for each gene type, were combined using commands detailed in `combine_distinct_seqs.sh` and the output was used to run `rss_finder.py`. The `rss_finder.py` script produces seven files, one for each chain-specific gene segment type. `rss_finder.py` requires a pickle file of known RSSs. I procured my own in the same format as [IgDetective](https://github.com/Immunotools/IgDetective)'s "motif" file, but with added, non-redundant RSS sequences from diverse species including those reported by [Ramsden *et al.* 1994](https://pubmed.ncbi.nlm.nih.gov/8208601/) (doi: 10.1093/nar/22.10.1785) and by the makers of [RSSsite](https://www.itb.cnr.it/rss/extratdata.html). This step could be improved in future iterations by also including RSSs from more species provided by [IMGT](https://www.imgt.org/genedb/).

3. <ins>**Finding adjacent and internal motifs for the three gene types:**</ins> The three V gene result files from the previous step were passed to `v_analysis.py`, the one IGHD gene result file was passed to `d_analysis.py`, and the three J gene files were passed to `j_analysis.py`. Each script does the following:
- `v_analysis.py`: Annotates sequences by finding V1 & V2 leader sequences (powered by consensus sequences taken from known [IMGT ferret leader sequences](https://www.imgt.org/genedb/) and a local version [SignalP-5.0](https://services.healthtech.dtu.dk/services/SignalP-5.0/)), and framework regions + complementarity determining regions (powered by [IgBLAST](https://ncbi.github.io/igblast/cook/How-to-set-up.html) with a reference database of a related species or, in this case, [IMGT-reported ferret sequences](https://www.imgt.org/vquest/refseqh.html#:~:text=IG%20%22V-REGION%22%2C%20%22D-REGION%22%2C%20%22J-REGION%22%2C%20%22C-GENE%20exon%22%20sets)). 
- `d_analysis.py`: Annotates sequences by finding translation frames
- `j_analysis.py`: Annotates sequences by finding J-motifs and necessary downstream splice donor

4. <ins>**Using Digger for start/stop positions:**</ins> In future iterations of our pipeline, this will hopefully not be necessary. But for expediency in this fist flagship ferret analysis, I turned to Digger to acquire precise start & stop positions of the aforementioned gene features (e.g. leader sequences, actual coding frames). To do this, position weighted matrixes were created using previously reported ferret sequences in accordance with the [Digger documentation](https://williamdlees.github.io/digger/_build/html/examples/rhesus_igh.html). Digger was applied in forward and reverse orientations to the ferret genome assembly, and the results take the form of three TSV files, one for each Ig chain type (heavy, kappa, lambda).

5. <ins>**Merging Digger results with our in-house results:**</ins> Unfortunately, the results from Digger suggest thousands of each gene type (e.g. thousands of IGHJ genes) which is exceedingly unlikely to be biologically real. But the purpose of Digger in our workflow is not gene discovery *per se*, but to use its inevitably overlapping results to help annotate our candidate gene loci. So, Digger annotations were merged with ours using the `combine_digger_genig.py` script which takes the aforementioned annottaed files from stsep B3 and Digger output files as input. The `format_overlapping_results.py` script was used to combine all `combine_digger_genig.py` output files and reformat them for the next step.

## C. Assigning functionality labels to gene candidates.
1. <ins>**Assigning genome-based functionality labels (functional, ORF, pseudogene) using IMGT's criteria:**</ins> The `assign_functionality.py` script was used to apply [IMGT's functionality criteria](https://pubmed.ncbi.nlm.nih.gov/20433708/) (doi: 10.1186/1471-2105-11-223) to annotated gene candidates from steps A & B.
2. <ins>**Making FASTAs for downstream transcriptomic analysis:**</ins> From the resulting TSV file, gene nucleotide sequences and identities (names) were extracted to make a FASTA file using `make_ig_gene_fastas.py`
3. <ins>**Rounding out discovered loci**</ins> Finally, in case this overarching BLAST->annotate->assign approach did not catch some trailing previously-discovered loci, the FASTA sequences from step C2 were supplemented with [IMGT's ferret sequences](https://www.imgt.org/vquest/refseqh.html#:~:text=IG%20%22V-REGION%22%2C%20%22D-REGION%22%2C%20%22J-REGION%22%2C%20%22C-GENE%20exon%22%20sets) using `supplement_imgt_seqs.py`
